File: src/adk_fluent/_harness/_streaming.py
"""Streaming bash — PTY-based shell with real-time output.

Unlike the blocking ``bash`` tool that waits for completion, the streaming
bash tool yields output chunks as they arrive. This is essential for
long-running commands (builds, tests, installs) where the LLM or user
needs progressive feedback.

Usage::

    streamer = StreamingBash(sandbox)
    async for chunk in streamer.run("npm test"):
        print(chunk, end="")

    # Or use as a tool that returns full output but streams to a callback
    bash_tool = make_streaming_bash(sandbox, on_output=print)
"""

from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator, Callable

from adk_fluent._harness._sandbox import SandboxPolicy

__all__ = ["StreamingBash", "make_streaming_bash"]


class StreamingBash:
    """PTY-based streaming shell execution.

    Uses ``asyncio.create_subprocess_shell`` with pipe-based streaming.
    Falls back to blocking subprocess if async loop is not available.
    """

    def __init__(self, sandbox: SandboxPolicy) -> None:
        self.sandbox = sandbox

    async def run(
        self,
        command: str,
        *,
        timeout: int = 120,
        on_output: Callable[[str], None] | None = None,
    ) -> AsyncIterator[str]:
        """Execute a command and yield output chunks as they arrive.

        Args:
            command: Shell command to execute.
            timeout: Maximum execution time in seconds.
            on_output: Optional callback for each output chunk.

        Yields:
            Output chunks as strings.
        """
        if not self.sandbox.allow_shell:
            yield "Error: shell execution is disabled by sandbox policy."
            return

        cwd = self.sandbox.workspace or None
        try:
            proc = await asyncio.create_subprocess_shell(
                command,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.STDOUT,
                cwd=cwd,
            )

            total_bytes = 0
            max_bytes = self.sandbox.max_output_bytes

            async def _read_with_timeout():
                try:
                    return await asyncio.wait_for(
                        proc.stdout.read(4096),
                        timeout=timeout,
                    )
                except TimeoutError:
                    proc.kill()
                    return None

            while True:
                data = await _read_with_timeout()
                if data is None:
                    yield f"\nError: command timed out after {timeout}s"
                    break
                if not data:
                    break

                chunk = data.decode("utf-8", errors="replace")
                total_bytes += len(data)

                if total_bytes > max_bytes:
                    yield f"\n... (truncated to {max_bytes} bytes)"
                    proc.kill()
                    break

                if on_output:
                    on_output(chunk)
                yield chunk

            await proc.wait()
            if proc.returncode and proc.returncode != 0:
                yield f"\nExit code: {proc.returncode}"

        except Exception as e:
            yield f"Error executing command: {e}"

    async def run_collected(
        self,
        command: str,
        *,
        timeout: int = 120,
        on_output: Callable[[str], None] | None = None,
    ) -> str:
        """Execute a command and return the full output as a string.

        Like ``run()`` but collects all chunks. If ``on_output`` is provided,
        it also streams chunks to the callback as they arrive.
        """
        parts: list[str] = []
        async for chunk in self.run(command, timeout=timeout, on_output=on_output):
            parts.append(chunk)
        return "".join(parts)


def make_streaming_bash(
    sandbox: SandboxPolicy,
    on_output: Callable[[str], None] | None = None,
) -> Callable:
    """Create a streaming-aware bash tool.

    The tool itself is synchronous (ADK requirement) but internally uses
    async streaming. If an ``on_output`` callback is provided, output
    chunks are streamed to it as they arrive.

    Args:
        sandbox: Sandbox policy.
        on_output: Optional callback for real-time output chunks.
    """
    streamer = StreamingBash(sandbox)

    def bash(command: str, timeout: int = 120) -> str:
        """Execute a shell command with streaming output.

        Args:
            command: The shell command to execute.
            timeout: Maximum execution time in seconds (default: 120).
        """
        if not sandbox.allow_shell:
            return "Error: shell execution is disabled by sandbox policy."

        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            loop = None

        if loop and loop.is_running():
            # We're inside an async context — use a thread
            import concurrent.futures

            with concurrent.futures.ThreadPoolExecutor(max_workers=1) as pool:
                future = pool.submit(
                    asyncio.run,
                    streamer.run_collected(command, timeout=timeout, on_output=on_output),
                )
                return future.result(timeout=timeout + 5)
        else:
            return asyncio.run(streamer.run_collected(command, timeout=timeout, on_output=on_output))

    return bash

File: ts/src/namespaces/harness/streaming.ts
/**
 * Streaming bash — async-iterator shell with real-time output chunks.
 *
 * The TypeScript port of `python/src/adk_fluent/_harness/_streaming.py`.
 *
 * Unlike the blocking process tools, `StreamingBash` yields output as it
 * arrives, which is essential for long-running commands (builds, tests,
 * installs) where the LLM or user wants progressive feedback.
 *
 * Usage:
 * ```ts
 * const streamer = new StreamingBash(sandbox);
 * for await (const chunk of streamer.run("npm test")) {
 *   process.stdout.write(chunk);
 * }
 * ```
 */

import { spawn } from "node:child_process";
import { type SandboxPolicy } from "./sandbox.js";

export interface StreamingBashRunOptions {
  /** Maximum execution time in seconds. Defaults to 120. */
  timeoutSec?: number;
  /** Optional callback fired for each output chunk before it is yielded. */
  onOutput?: (chunk: string) => void;
}

/**
 * Streaming shell executor.
 *
 * Spawns commands via `bash -c <command>` inside the sandbox workspace,
 * yielding stdout+stderr chunks as they arrive. Honors `sandbox.allowShell`,
 * `sandbox.workspace`, and `sandbox.maxOutputBytes` exactly like the
 * blocking `processTools()` flow.
 */
export class StreamingBash {
  constructor(public readonly sandbox: SandboxPolicy) {}

  /**
   * Execute a command and yield output chunks as they arrive.
   *
   * Output is decoded as UTF-8. Stderr is merged into stdout. The iterator
   * completes when the process exits, the timeout fires, or the cumulative
   * output exceeds `sandbox.maxOutputBytes`.
   */
  async *run(command: string, opts: StreamingBashRunOptions = {}): AsyncIterableIterator<string> {
    if (!this.sandbox.allowShell) {
      yield "Error: shell execution is disabled by sandbox policy.";
      return;
    }

    const timeoutSec = opts.timeoutSec ?? 120;
    const cwd = this.sandbox.workspace ?? process.cwd();
    const maxBytes = this.sandbox.maxOutputBytes;

    const child = spawn("bash", ["-c", command], { cwd });

    // Buffered chunk queue. The producer (stdout/stderr listeners and the
    // exit/error handlers) push items; the consumer (this generator) drains
    // them via the resolver-promise pattern below.
    type Item = { kind: "chunk"; value: string } | { kind: "end"; value?: string };
    const queue: Item[] = [];
    let resolveNext: (() => void) | null = null;
    const wake = () => {
      if (resolveNext) {
        const r = resolveNext;
        resolveNext = null;
        r();
      }
    };

    let totalBytes = 0;
    let killed = false;

    const handleData = (chunk: Buffer) => {
      if (killed) return;
      let text = chunk.toString("utf8");
      totalBytes += chunk.length;
      if (totalBytes > maxBytes) {
        text += `\n... (truncated to ${maxBytes} bytes)`;
        queue.push({ kind: "chunk", value: text });
        queue.push({ kind: "end" });
        killed = true;
        child.kill("SIGTERM");
        wake();
        return;
      }
      queue.push({ kind: "chunk", value: text });
      wake();
    };

    child.stdout?.on("data", handleData);
    child.stderr?.on("data", handleData);

    child.on("error", (err) => {
      queue.push({ kind: "end", value: `Error executing command: ${err.message}` });
      wake();
    });

    child.on("close", (code, signal) => {
      const trailing =
        signal && !killed
          ? `\nProcess terminated by ${signal}`
          : code != null && code !== 0
            ? `\nExit code: ${code}`
            : undefined;
      queue.push({ kind: "end", value: trailing });
      wake();
    });

    const timer = setTimeout(() => {
      if (!killed) {
        killed = true;
        child.kill("SIGKILL");
        queue.push({ kind: "end", value: `\nError: command timed out after ${timeoutSec}s` });
        wake();
      }
    }, timeoutSec * 1000);
    // Don't keep the event loop alive on the timeout alone.
    timer.unref?.();

    try {
      while (true) {
        if (queue.length === 0) {
          await new Promise<void>((resolve) => {
            resolveNext = resolve;
          });
        }
        const item = queue.shift();
        if (!item) continue;
        if (item.kind === "end") {
          if (item.value) {
            if (opts.onOutput) opts.onOutput(item.value);
            yield item.value;
          }
          return;
        }
        if (opts.onOutput) opts.onOutput(item.value);
        yield item.value;
      }
    } finally {
      clearTimeout(timer);
      if (!killed && child.exitCode == null) {
        try {
          child.kill("SIGTERM");
        } catch {
          // process already gone
        }
      }
    }
  }

  /**
   * Execute a command and return the full collected output as a string.
   *
   * Like `run()` but joins every chunk. If `onOutput` is supplied it still
   * fires for each chunk as it arrives, so callers can stream and collect
   * simultaneously.
   */
  async runCollected(command: string, opts: StreamingBashRunOptions = {}): Promise<string> {
    const parts: string[] = [];
    for await (const chunk of this.run(command, opts)) {
      parts.push(chunk);
    }
    return parts.join("");
  }
}
